File: examples/php-proxy/proxy.php
<?php
/**
 * GoClaw WebChat Proxy Server — PHP (Ratchet + ReactPHP)
 *
 * Lightweight WebSocket proxy that sits between the chat widget and GoClaw Gateway.
 * The proxy injects the auth token server-side so it never reaches the browser.
 *
 * Usage:
 *   composer install
 *   cp .env.example .env   # fill in GOCLAW_URL and GOCLAW_TOKEN
 *   php proxy.php
 *
 * Environment variables:
 *   GOCLAW_URL       — Gateway WebSocket URL (required, e.g. "ws://localhost:9090/ws")
 *   GOCLAW_TOKEN     — Gateway auth token (required, kept server-side)
 *   PORT             — Proxy listen port (default: 3100)
 *   ALLOWED_ORIGINS  — Comma-separated origin allowlist (empty = allow all)
 *   PROXY_API_KEY    — Optional API key for proxy authentication
 *   DEFAULT_AGENT_ID — Default agent ID injected into chat.send if client omits it
 */

require __DIR__ . '/vendor/autoload.php';

use Ratchet\MessageComponentInterface;
use Ratchet\ConnectionInterface;
use Ratchet\Http\HttpServer;
use Ratchet\Server\IoServer;
use Ratchet\WebSocket\WsServer;
use Ratchet\Http\OriginCheck;
use React\EventLoop\Loop;
use React\Socket\SocketServer;

// ── Load .env ───────────────────────────────────────────────────────────────

if (class_exists('Dotenv\Dotenv') && file_exists(__DIR__ . '/.env')) {
    $dotenv = Dotenv\Dotenv::createImmutable(__DIR__);
    $dotenv->load();
}

// ── Config ──────────────────────────────────────────────────────────────────

$GOCLAW_URL      = $_ENV['GOCLAW_URL'] ?? getenv('GOCLAW_URL') ?: '';
$GOCLAW_TOKEN    = $_ENV['GOCLAW_TOKEN'] ?? getenv('GOCLAW_TOKEN') ?: '';
$PORT            = (int)($_ENV['PORT'] ?? getenv('PORT') ?: '3100');
$PROXY_API_KEY   = $_ENV['PROXY_API_KEY'] ?? getenv('PROXY_API_KEY') ?: '';
$DEFAULT_AGENT_ID = $_ENV['DEFAULT_AGENT_ID'] ?? getenv('DEFAULT_AGENT_ID') ?: '';

$originsRaw = $_ENV['ALLOWED_ORIGINS'] ?? getenv('ALLOWED_ORIGINS') ?: '';
$ALLOWED_ORIGINS = array_filter(array_map('trim', explode(',', $originsRaw)));

if (empty($GOCLAW_URL)) {
    fwrite(STDERR, "ERROR: GOCLAW_URL environment variable is required\n");
    exit(1);
}

if (empty($GOCLAW_TOKEN)) {
    fwrite(STDERR, "WARNING: GOCLAW_TOKEN not set — proxy will connect without authentication\n");
}

// ── Proxy Component ─────────────────────────────────────────────────────────

class GoclawProxy implements MessageComponentInterface
{
    private string $goclawUrl;
    private string $goclawToken;
    private string $proxyApiKey;
    private string $defaultAgentId;
    private array $allowedOrigins;
    private int $activeConnections = 0;

    /** @var \SplObjectStorage<ConnectionInterface, \Ratchet\Client\WebSocket|null> */
    private \SplObjectStorage $upstreams;

    public function __construct(
        string $goclawUrl,
        string $goclawToken,
        string $proxyApiKey,
        string $defaultAgentId,
        array  $allowedOrigins,
    ) {
        $this->goclawUrl      = $goclawUrl;
        $this->goclawToken    = $goclawToken;
        $this->proxyApiKey    = $proxyApiKey;
        $this->defaultAgentId = $defaultAgentId;
        $this->allowedOrigins = $allowedOrigins;
        $this->upstreams      = new \SplObjectStorage();
    }

    public function getActiveConnections(): int
    {
        return $this->activeConnections;
    }

    public function onOpen(ConnectionInterface $conn): void
    {
        // Check API key if configured
        if (!$this->checkApiKey($conn)) {
            echo "[proxy] invalid or missing API key\n";
            $conn->close();
            return;
        }

        // Check origin if allowlist is configured
        if (!$this->checkOrigin($conn)) {
            echo "[proxy] origin rejected\n";
            $conn->close();
            return;
        }

        $this->activeConnections++;
        echo "[proxy] client connected (active={$this->activeConnections})\n";

        // Connect to upstream GoClaw Gateway
        $this->upstreams[$conn] = null; // placeholder until connected

        $connector = new \Ratchet\Client\Connector(Loop::get());
        $connector($this->goclawUrl)->then(
            function (\Ratchet\Client\WebSocket $upstream) use ($conn) {
                if (!$this->upstreams->contains($conn)) {
                    // Client already disconnected
                    $upstream->close();
                    return;
                }

                $this->upstreams[$conn] = $upstream;
                echo "[proxy] upstream connected\n";

                // Relay upstream -> client
                $upstream->on('message', function ($msg) use ($conn) {
                    $sanitized = $this->sanitizeUpstreamFrame((string)$msg);
                    $conn->send($sanitized);
                });

                $upstream->on('close', function () use ($conn) {
                    $conn->close();
                });
            },
            function (\Exception $e) use ($conn) {
                echo "[proxy] upstream connection failed: {$e->getMessage()}\n";
                $conn->close();
            }
        );
    }

    public function onMessage(ConnectionInterface $conn, $msg): void
    {
        $upstream = $this->upstreams[$conn] ?? null;
        if ($upstream === null) {
            return; // upstream not connected yet, drop message
        }

        $modified = $this->interceptFrame((string)$msg);
        $upstream->send($modified);
    }

    public function onClose(ConnectionInterface $conn): void
    {
        $this->activeConnections--;
        echo "[proxy] client disconnected (active={$this->activeConnections})\n";

        if ($this->upstreams->contains($conn)) {
            $upstream = $this->upstreams[$conn];
            if ($upstream !== null) {
                $upstream->close();
            }
            $this->upstreams->detach($conn);
        }
    }

    public function onError(ConnectionInterface $conn, \Exception $e): void
    {
        echo "[proxy] error: {$e->getMessage()}\n";
        $conn->close();
    }

    // ── Frame interception ──────────────────────────────────────────────────

    private function interceptFrame(string $raw): string
    {
        $frame = json_decode($raw, true);
        if (!is_array($frame) || ($frame['type'] ?? '') !== 'req') {
            return $raw;
        }

        $modified = false;

        // Inject gateway token into connect frame
        if (($frame['method'] ?? '') === 'connect' && $this->goclawToken !== '') {
            $frame['params'] = $frame['params'] ?? [];
            $frame['params']['token'] = $this->goclawToken;
            $modified = true;
        }

        // Inject default agentId into chat.send if not set by client
        if (
            ($frame['method'] ?? '') === 'chat.send'
            && $this->defaultAgentId !== ''
            && empty($frame['params']['agentId'])
        ) {
            $frame['params'] = $frame['params'] ?? [];
            $frame['params']['agentId'] = $this->defaultAgentId;
            $modified = true;
        }

        return $modified ? json_encode($frame, JSON_UNESCAPED_SLASHES) : $raw;
    }

    private function sanitizeUpstreamFrame(string $raw): string
    {
        $frame = json_decode($raw, true);
        if (
            is_array($frame)
            && ($frame['type'] ?? '') === 'res'
            && isset($frame['payload']['token'])
        ) {
            unset($frame['payload']['token']);
            return json_encode($frame, JSON_UNESCAPED_SLASHES);
        }
        return $raw;
    }

    // ── Auth helpers ────────────────────────────────────────────────────────

    private function checkApiKey(ConnectionInterface $conn): bool
    {
        if ($this->proxyApiKey === '') {
            return true;
        }

        $request = $conn->httpRequest ?? null;
        if ($request === null) {
            return false;
        }

        // Check query param: ?apiKey=xxx
        $query = [];
        parse_str($request->getUri()->getQuery(), $query);
        if (($query['apiKey'] ?? '') === $this->proxyApiKey) {
            return true;
        }

        // Check header: X-API-Key
        $headerKey = $request->getHeaderLine('X-API-Key');
        return $headerKey === $this->proxyApiKey;
    }

    private function checkOrigin(ConnectionInterface $conn): bool
    {
        if (empty($this->allowedOrigins)) {
            return true;
        }

        $request = $conn->httpRequest ?? null;
        if ($request === null) {
            return false;
        }

        $origin = $request->getHeaderLine('Origin');
        if ($origin === '') {
            return false; // reject missing origin when allowlist is active
        }

        return in_array('*', $this->allowedOrigins) || in_array($origin, $this->allowedOrigins);
    }
}

// ── Server setup ────────────────────────────────────────────────────────────

$loop = Loop::get();

$proxy = new GoclawProxy(
    $GOCLAW_URL,
    $GOCLAW_TOKEN,
    $PROXY_API_KEY,
    $DEFAULT_AGENT_ID,
    $ALLOWED_ORIGINS,
);

$wsServer = new WsServer($proxy);
$wsServer->enableKeepAlive($loop, 30);

// Health check + WebSocket on same port using custom HTTP handler
$httpServer = new HttpServer($wsServer);

$socket = new SocketServer("0.0.0.0:{$PORT}", [], $loop);

$server = new IoServer($httpServer, $socket, $loop);

echo "[proxy] listening on :{$PORT}\n";
echo "[proxy] upstream: {$GOCLAW_URL}\n";
echo "[proxy] auth token: " . ($GOCLAW_TOKEN ? 'configured' : 'NOT SET') . "\n";
echo "[proxy] API key: " . ($PROXY_API_KEY ? 'required' : 'disabled') . "\n";
if (!empty($ALLOWED_ORIGINS)) {
    echo "[proxy] allowed origins: " . implode(', ', $ALLOWED_ORIGINS) . "\n";
} else {
    echo "[proxy] allowed origins: * (all)\n";
}

$server->run();

File: examples/python-proxy/proxy.py
"""
GoClaw WebChat Proxy Server — Python (aiohttp)

Lightweight WebSocket proxy that sits between the chat widget and GoClaw Gateway.
The proxy injects the auth token server-side so it never reaches the browser.

Usage:
  pip install -r requirements.txt
  cp .env.example .env   # fill in GOCLAW_URL and GOCLAW_TOKEN
  python proxy.py

Environment variables:
  GOCLAW_URL       — Gateway WebSocket URL (required, e.g. "ws://localhost:9090/ws")
  GOCLAW_TOKEN     — Gateway auth token (required, kept server-side)
  PORT             — Proxy listen port (default: 3100)
  ALLOWED_ORIGINS  — Comma-separated origin allowlist (empty = allow all)
  PROXY_API_KEY    — Optional API key for proxy authentication
  DEFAULT_AGENT_ID — Default agent ID injected into chat.send if client omits it
"""

import asyncio
import json
import os
import signal
import sys

import aiohttp
from aiohttp import web

# ── Config ──────────────────────────────────────────────────────────────────

GOCLAW_URL = os.environ.get("GOCLAW_URL", "")
GOCLAW_TOKEN = os.environ.get("GOCLAW_TOKEN", "")
PORT = int(os.environ.get("PORT", "3100"))
ALLOWED_ORIGINS = [
    o.strip()
    for o in os.environ.get("ALLOWED_ORIGINS", "").split(",")
    if o.strip()
]
PROXY_API_KEY = os.environ.get("PROXY_API_KEY", "")
DEFAULT_AGENT_ID = os.environ.get("DEFAULT_AGENT_ID", "")

if not GOCLAW_URL:
    print("ERROR: GOCLAW_URL environment variable is required")
    sys.exit(1)

if not GOCLAW_TOKEN:
    print("WARNING: GOCLAW_TOKEN not set — proxy will connect without authentication")


# ── Helpers ─────────────────────────────────────────────────────────────────

def check_origin(request: web.Request) -> bool:
    """Validate request origin against allowlist. Empty list = allow all."""
    if not ALLOWED_ORIGINS:
        return True
    origin = request.headers.get("Origin", "")
    if not origin:
        return False  # reject missing origin when allowlist is active
    return "*" in ALLOWED_ORIGINS or origin in ALLOWED_ORIGINS


def check_api_key(request: web.Request) -> bool:
    """Validate API key from query param or header. No key configured = allow all."""
    if not PROXY_API_KEY:
        return True
    key = request.query.get("apiKey") or request.headers.get("X-API-Key", "")
    return key == PROXY_API_KEY


def intercept_frame(raw: str) -> str:
    """Intercept client frames: inject token into connect, default agentId into chat.send."""
    try:
        frame = json.loads(raw)
    except (json.JSONDecodeError, TypeError):
        return raw

    if frame.get("type") != "req":
        return raw

    modified = False

    # Inject gateway token into connect frame
    if frame.get("method") == "connect" and GOCLAW_TOKEN:
        frame.setdefault("params", {})["token"] = GOCLAW_TOKEN
        modified = True

    # Inject default agentId into chat.send if not set by client
    if (
        frame.get("method") == "chat.send"
        and DEFAULT_AGENT_ID
        and not frame.get("params", {}).get("agentId")
    ):
        frame.setdefault("params", {})["agentId"] = DEFAULT_AGENT_ID
        modified = True

    return json.dumps(frame) if modified else raw


def sanitize_upstream_frame(raw: str) -> str:
    """Strip token fields from upstream responses (defense in depth)."""
    try:
        frame = json.loads(raw)
        if frame.get("type") == "res" and "token" in frame.get("payload", {}):
            del frame["payload"]["token"]
            return json.dumps(frame)
    except (json.JSONDecodeError, TypeError):
        pass
    return raw


# ── WebSocket proxy handler ────────────────────────────────────────────────

active_connections = 0


async def ws_proxy(request: web.Request) -> web.WebSocketResponse:
    """Handle a single WebSocket proxy session: client <-> upstream."""
    global active_connections

    if not check_origin(request):
        return web.Response(status=403, text="Origin not allowed")

    if not check_api_key(request):
        return web.Response(status=401, text="Unauthorized")

    client_ws = web.WebSocketResponse(max_msg_size=512 * 1024)
    await client_ws.prepare(request)

    active_connections += 1
    print(f"[proxy] client connected (active={active_connections})")

    # Connect to upstream GoClaw Gateway
    session = aiohttp.ClientSession()
    try:
        upstream_ws = await session.ws_connect(GOCLAW_URL, max_msg_size=512 * 1024)
    except Exception as exc:
        print(f"[proxy] upstream connection failed: {exc}")
        active_connections -= 1
        await session.close()
        await client_ws.close(code=1011, message=b"upstream connection failed")
        return client_ws

    print("[proxy] upstream connected")

    async def relay_upstream_to_client() -> None:
        """Forward upstream messages to client, stripping token fields."""
        try:
            async for msg in upstream_ws:
                if msg.type == aiohttp.WSMsgType.TEXT:
                    await client_ws.send_str(sanitize_upstream_frame(msg.data))
                elif msg.type in (aiohttp.WSMsgType.CLOSED, aiohttp.WSMsgType.ERROR):
                    break
        except Exception:
            pass
        finally:
            if not client_ws.closed:
                await client_ws.close()

    # Start upstream -> client relay in background
    relay_task = asyncio.create_task(relay_upstream_to_client())

    # Client -> upstream relay (main loop)
    try:
        async for msg in client_ws:
            if msg.type == aiohttp.WSMsgType.TEXT:
                modified = intercept_frame(msg.data)
                await upstream_ws.send_str(modified)
            elif msg.type in (aiohttp.WSMsgType.CLOSED, aiohttp.WSMsgType.ERROR):
                break
    except Exception:
        pass
    finally:
        active_connections -= 1
        print(f"[proxy] client disconnected (active={active_connections})")
        relay_task.cancel()
        if not upstream_ws.closed:
            await upstream_ws.close()
        await session.close()

    return client_ws


# ── Health check ────────────────────────────────────────────────────────────

async def health(_request: web.Request) -> web.Response:
    return web.json_response({"status": "ok", "connections": active_connections})


# ── App setup ───────────────────────────────────────────────────────────────

app = web.Application()
app.router.add_get("/ws", ws_proxy)
app.router.add_get("/health", health)

if __name__ == "__main__":
    # Load .env file if python-dotenv is available
    try:
        from dotenv import load_dotenv
        load_dotenv()
    except ImportError:
        pass

    print(f"[proxy] listening on :{PORT}")
    print(f"[proxy] upstream: {GOCLAW_URL}")
    print(f"[proxy] auth token: {'configured' if GOCLAW_TOKEN else 'NOT SET'}")
    print(f"[proxy] API key: {'required' if PROXY_API_KEY else 'disabled'}")
    if ALLOWED_ORIGINS:
        print(f"[proxy] allowed origins: {', '.join(ALLOWED_ORIGINS)}")
    else:
        print("[proxy] allowed origins: * (all)")

    web.run_app(app, port=PORT, print=None)
